### PR TITLE
[flytepropeller] Support attribute access on promise

### PR DESCRIPTION
**Pull request by ByronHsu**
## Tracking issue

flyteorg/flyte#3864

## Describe your changes

See https://github.com/flyteorg/flyteidl/pull/439#issuecomment-1763223435

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [ ] All commits are signed-off.

### CODE REVIEW

File: flytepropeller/pkg/compiler/validators/bindings.go
Comment (hamersaw) [inline flytepropeller/pkg/compiler/validators/bindings.go:128]: these aren't used outside of the for loop are they? are they necessary?

File: flytepropeller/pkg/controller/nodes/attr_path_resolver_test.go
Comment (hamersaw) [inline flytepropeller/pkg/controller/nodes/attr_path_resolver_test.go:43]: does it make sense to have these comments next to the test cases?

**Review by hamersaw — CHANGES_REQUESTED** [review]
A few nit comments. Otherwise this looks great! This is going to be so powerful.

**Review by hamersaw — APPROVED** [review]
Looks great! Will wait for one more approve before merging to try and stick with old idl requirements.

File: flytepropeller/pkg/controller/nodes/attr_path_resolver.go
Comment (wild-endeavor) [inline flytepropeller/pkg/controller/nodes/attr_path_resolver.go:94]: just one question.  if you just have a typing.Dict[str, DataClass], and you bind to one of the dataclasses.  you end up in this condition right?

What is this line for?  `newSt, err := structpb.NewStruct(obj)`

In fact in this case - line 22 above:
`tmpVal, exist = currVal.GetMap().GetLiterals()[attr.GetStringValue()]`
won't tmpVal already be a Literal?  can't you just return it?  it won't get returned right now, it'll fall into the condition on line 41.

Comment (ByronHsu) [reply flytepropeller/pkg/controller/nodes/attr_path_resolver.go:94]: i guess you are talking about the case that where `bindAttrPath[count:]` is 0 and the `currVal` on line41 is a datatclass. Currently we convert convert struct to map, and convert it back to struct without doing other things because line 63 the loop will be skipped. 

adding a check about `len(bindAttrPath[count:]) == 0` on line 41 can circumvent the unnecessary work, but I think it's ok to keep it as-is because it should not be a large overhead.

**Review by wild-endeavor — APPROVED** [review]

**Pull request merged**
